components/fxa-client/src/oauth.rs

```diff
@@ -871,6 +914,57 @@ mod tests {
         assert_eq!(auth_status.active, true);
     }
 
+    #[test]
+    fn test_check_authorization_status_circuit_breaker() {
+        let config = Config::stable_dev("12345678", "https://foo.bar");
+        let mut fxa = FirefoxAccount::with_config(config);
+
+        let refresh_token_scopes = std::collections::HashSet::new();
+        fxa.state.refresh_token = Some(RefreshToken {
+            token: "refresh_token".to_owned(),
+            scopes: refresh_token_scopes,
+        });
+
+        let mut client = FxAClientMock::new();
+        // This copy-pasta (equivalent to `.returns(..).times(5)`) is there
+        // because `Error` is not cloneable :/
+        client
+            .expect_oauth_introspect_refresh_token(mockiato::Argument::any, |token| {
+                token.partial_eq("refresh_token")
+            })
+            .returns_once(Ok(IntrospectResponse { active: true }));
+        client
+            .expect_oauth_introspect_refresh_token(mockiato::Argument::any, |token| {
+                token.partial_eq("refresh_token")
+            })
+            .returns_once(Ok(IntrospectResponse { active: true }));
+        client
+            .expect_oauth_introspect_refresh_token(mockiato::Argument::any, |token| {
+                token.partial_eq("refresh_token")
+            })
+            .returns_once(Ok(IntrospectResponse { active: true }));
+        client
+            .expect_oauth_introspect_refresh_token(mockiato::Argument::any, |token| {
+                token.partial_eq("refresh_token")
+            })
+            .returns_once(Ok(IntrospectResponse { active: true }));
+        client
+            .expect_oauth_introspect_refresh_token(mockiato::Argument::any, |token| {
+                token.partial_eq("refresh_token")
+            })
+            .returns_once(Ok(IntrospectResponse { active: true }));
+        client.expect_oauth_introspect_refresh_token_calls_in_order();
+        fxa.set_client(Arc::new(client));
+
+        for _ in 0..5 {
+            assert!(fxa.check_authorization_status().is_ok());
+        }
+        match fxa.check_authorization_status() {
+            Ok(_) => unreachable!("should not happen"),
+            Err(err) => assert!(matches!(err.kind(), ErrorKind::AuthCircuitBreakerError)),
+        }
+    }
```

Comment:
Ideally, I think we'd also test that the access is allowed again after more tokens accumulate in the bucket. Are we set up to mock `util::now` in order to test this easily?

Reply:
Also you wouldn't necessary have to mock out all of the request infra to this this, just make a more focused unit-test on the behaviour of `check()`.